engine/allocator.hpp

```diff
@@ -26,6 +37,277 @@ struct SpaceEntry {
+   * This helps to avoid possible bottleneck when multiple threads
+   * access the std::atomic `global_allocated_size_`.
+   *
+   * @param max_access_threads
+   * @return true Success
+   * @return false Failure
+   */
+  bool EnableThreadLocalCounters(uint32_t max_access_threads) {
+    if (max_access_threads <= 1) {
+      return true;
+    }
+
+    // Should not be enabled more than once
+    if (!thread_local_counter_enabled_) {
+      std::lock_guard<std::mutex> lg(allocator_mu_);
+      if (!thread_local_counter_enabled_) {
+        thread_local_counter_enabled_ = true;
+        thread_allocated_sizes_.resize(max_access_threads);
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  void LogAllocation(int tid, size_t sz) {
+    if (tid == -1 || !thread_local_counter_enabled_) {
+      global_allocated_size_.fetch_add(sz);
+    } else {
+      assert(tid >= 0);
+      thread_allocated_sizes_[tid] += sz;
+    }
+  }
+
+  void LogDeallocation(int tid, size_t sz) {
+    if (tid == -1 || !thread_local_counter_enabled_) {
+      global_allocated_size_.fetch_sub(sz);
+    } else {
+      assert(tid >= 0);
+      thread_allocated_sizes_[tid] -= sz;
+    }
+  }
+
+  std::int64_t BytesAllocated() const {
+    std::int64_t total = 0;
+    for (auto const& thread_size : thread_allocated_sizes_) {
+      total += thread_size;
+    }
+    total += global_allocated_size_.load();
+    return total;
+  }
+
+  // translate offset of allocated space to address
+  inline void* offset2addr_checked(MemoryOffsetType offset) const {
+    assert(validate_offset(offset) && "Trying to access invalid offset");
+    return base_addr_ + offset;
+  }
+
+  inline void* offset2addr(MemoryOffsetType offset) const {
+    if (validate_offset(offset)) {
+      return base_addr_ + offset;
+    }
+    return nullptr;
+  }
+
+  template <typename T>
+  inline T* offset2addr_checked(MemoryOffsetType offset) const {
+    return static_cast<T*>(offset2addr_checked(offset));
+  }
+
+  template <typename T>
+  inline T* offset2addr(MemoryOffsetType offset) const {
+    return static_cast<T*>(offset2addr(offset));
+  }
+
+  // translate address of allocated space to offset
+  inline MemoryOffsetType addr2offset_checked(const void* addr) const {
+    assert((char*)addr >= base_addr_);
+    MemoryOffsetType offset = (char*)addr - base_addr_;
+    assert(validate_offset(offset) && "Trying to create invalid offset");
+    return offset;
+  }
+
+  inline MemoryOffsetType addr2offset(const void* addr) const {
+    if (addr) {
+      MemoryOffsetType offset = (char*)addr - base_addr_;
+      if (validate_offset(offset)) {
+        return offset;
+      }
+    }
+    return kNullMemoryOffset;
+  }
+
+  inline bool validate_offset(uint64_t offset) const {
+    return offset < max_offset_;
+  }
+
+ private:
+  char* base_addr_;
+  uint64_t max_offset_;
+
+  bool thread_local_counter_enabled_{};
+  std::vector<std::int64_t> thread_allocated_sizes_;
```

Comment:
False sharing, potential bottleneck.

Reply:
I'd like to use `std::int64_t*` in next commit.